### PR TITLE
Implement ERC20Gateway::confirmWithdraw

### LINKED ISSUE #275: Implement ERC20Gateway::confirmWithdraw
Implement `confirmWithdraw` method in ERC20Gateway
- Transfer the ERC20 token from `ERC20Gateway` contract to the beneficiary address provided for the withdrawer.
- Merkle proof is required.
```
    ERC20Gateway.confirmWithdraw
        - utility token address
        - value token address
        - amount
        - beneficiary address
        - gas price
        - gas limit
        - sender
        - block number
        - storage proof  
```
- This method should also transfer reward/fee to the `msg.sender` address.

  Emit event: WithdrawIntentConfirmed
```
WithdrawIntentConfirmed
  - message hash  
```

### PR DESCRIPTION
**Pull request by Vindeep07**
Fixes #275 

- Implements ERC20Gateway::confirmWithdraw()

- Unit tests are not included because populating value token address is necessary in proof generation for ERC20Cogateway::withdraw(). While testing, valueToken is making a transaction. Hence, it must be a live contract. It will be efficient that it can be handled in integration tests.
Same approach has been followed for consensusCogateway::confirmWithdraw. 
cc: @deepesh-kn

### CODE REVIEW

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 1]
```diff
@@ -26,6 +27,11 @@ import "../proxies/MasterCopyNonUpgradable.sol";
  */
 contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
 
+    /** Events */
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
    /* Events */
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 2]
```diff
@@ -26,6 +27,11 @@ import "../proxies/MasterCopyNonUpgradable.sol";
...
+
+    event WithdrawIntentConfirmed ( bytes32 messageHash );
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
    event WithdrawIntentConfirmed(bytes32 messageHash);
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 3]
```diff
@@ -110,4 +116,111 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+    /**
+     * @notice Confirm withdraw in order to transfer value token.
+     *
+     * @param _utilityToken Address of utility token.
+     * @param _valueToken Address of value token.
+     * @param _amount Value token amount for withdrawal.
+     * @param _beneficiary Address of beneficiary where tokens will be withdrawn.
+     * @param _feeGasPrice Gas price at which fee will be calculated.
+     * @param _feeGasLimit Gas limit at which fee will be capped.
+     * @param _blockNumber Block number of Aux chain against which storage
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
     * @param _blockNumber Block number of auxiliary chain against which storage
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 4]
```diff
@@ -110,4 +116,111 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+                           proof is generated.
+     * @param _withdrawer Address of the withdrawer account.
+     * @param _rlpParentNodes Storage merkle proof to verify message declaration
+                              on the origin chain.
+     *
+     * @return messageHash_ Message Hash
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
     * @return messageHash_ Message hash.
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 5]
```diff
@@ -110,4 +116,111 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+        uint256 _blockNumber,
+        bytes calldata _rlpParentNodes
+    )
+        external
+        returns (bytes32 messageHash_)
+    {
+        uint256 initialGas = gasleft();
+        require(
+            _utilityToken != address(0),
+            "Utility Token address must not be 0."
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
            "Utility Token address is 0."
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 6]
```diff
@@ -110,4 +116,111 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+        );
+        require(
+            _valueToken != address(0),
+            "Value Token address must not be 0."
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
            "Value Token address is 0."
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 7]
```diff
@@ -110,4 +116,111 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+        );
+        require(
+            _amount != 0,
+            "Withdraw amount should be greater than 0."
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
            "Withdraw amount is 0."
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 8]
```diff
@@ -110,4 +116,111 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+        );
+        require(
+            _beneficiary != address(0),
+            "Beneficiary address must not be 0."
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
            "Beneficiary address is 0."
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 9]
```diff
@@ -110,4 +116,111 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+        );
+        require(
+            _withdrawer != address(0),
+            "Withdrawer address must not be 0."
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
            "Withdrawer address is 0."
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 10]
```diff
@@ -110,4 +116,111 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+     * @param _feeGasLimit Gas limit at which fee will be capped.
+     * @param _blockNumber Block number of Aux chain against which storage
+                           proof is generated.
+     * @param _withdrawer Address of the withdrawer account.
+     * @param _rlpParentNodes Storage merkle proof to verify message declaration
+                              on the origin chain.
+     *
+     * @return messageHash_ Message Hash
+     *
+     * \pre `_amount` is not 0.
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol:246]: Please update \pre conditions ordering as per method call ordering.
e.g. "Utility Token address is 0." should come first.

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 11]
```diff
@@ -110,4 +116,111 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+     *
+     * \post Adds a new entry in `inbox` mapping storage variable. The value is
+     *       set as `true` for `messagehash_` in `inbox` mapping. The
+     *       `messageHash_` is calculated by `MessageInbox.confirmMessage`.
+     * \post Update the nonces storage mapping variable by incrementing the
+     *       value for `msg.sender` by one.
+     * \post Transfers the `_valueToken` token to the `msg.sender` address as a fees.
+     *       The `fees` amount is calculated by `ERC20GatewayBase::reward()`
+     * \post Transfer the `_valueToken` token to the `_beneficiary` address. The
+     *       transfer amount calculated as `_amount-fees`
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
     *       transfer amount is calculated as `_amount-fees`.
```

Comment (abhayks1) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Add \post condition for event emit.

File: contracts/erc20-gateway/ERC20GatewayBase.sol [hunk 12]
```diff
@@ -39,6 +39,11 @@ contract ERC20GatewayBase {
         "WithdrawIntent(address valueToken,address utilityToken,uint256 amount,address beneficiary)"
     );
 
+    /* Storage */
+
+    /** Mapping of message sender and nonce. */
+    mapping(address => uint256) public  nonces;
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20GatewayBase.sol]:
```suggestion
    mapping(address => uint256) public nonces;
```

File: contracts/erc20-gateway/ERC20GatewayBase.sol [hunk 13]
```diff
@@ -39,6 +39,11 @@ contract ERC20GatewayBase {
         "WithdrawIntent(address valueToken,address utilityToken,uint256 amount,address beneficiary)"
     );
 
+    /* Storage */
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20GatewayBase.sol:43]: Add two new lines before top level declarations.
Do read the guideline: https://github.com/mosaicdao/developer-guidelines/blob/master/SOLIDITY_STYLE_GUIDE.md#blank-lines

**Review by abhayks1 — DISMISSED** [review]
Nice work 👍 
There are inline comments.

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 14]
```diff
@@ -110,4 +116,111 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
         emit GatewayProven(messageInbox, _blockNumber);
     }
 
+    /**
+     * @notice Confirm withdraw in order to transfer value token.
+     *
+     * @param _utilityToken Address of utility token.
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
     * @param _utilityToken Address of utility token contract.
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 15]
```diff
@@ -110,4 +116,111 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+     * @param _valueToken Address of value token.
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
     * @param _valueToken Address of value token contract.
```

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 1 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Done.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 2 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Fixed.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 3 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Changed.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 4 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Done.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 5 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Fixed.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 6 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Fixed.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 7 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Fixed.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 8 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Fixed.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 9 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Fixed.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 10 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Gateway.sol:246]: Updated.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 11 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Added \post condition for event emit.

File: contracts/erc20-gateway/ERC20GatewayBase.sol (hunk 12 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20GatewayBase.sol]: Fixed.

File: contracts/erc20-gateway/ERC20GatewayBase.sol (hunk 13 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20GatewayBase.sol:43]: Understood and changed.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 14 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Fixed.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 15 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Fixed.

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 16]
```diff
@@ -18,6 +18,7 @@ import "./ERC20GatewayBase.sol";
 import "../ERC20I.sol";
 import "../message-bus/MessageBus.sol";
 import "../proxies/MasterCopyNonUpgradable.sol";
+import "../ERC20I.sol";
```
Comment (deepesh-kn) [inline contracts/erc20-gateway/ERC20Gateway.sol]: This is already imported, please remove this line.

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 17]
```diff
@@ -219,4 +223,114 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+     * \pre `_beneficiary` address is not 0.
+     * \pre `_withdrawer` address is not 0.
+     * \pre `_rlpParentNodes` is not 0.
+     *
+     * \post Adds a new entry in `inbox` mapping storage variable. The value is
+     *       set as `true` for `messagehash_` in `inbox` mapping. The
+     *       `messageHash_` is calculated by `MessageInbox.confirmMessage`.
+     * \post Transfers the `_valueToken` token to the `msg.sender` address as a fees.
+     *       The `fees` amount is calculated by `ERC20GatewayBase::reward()`
+     * \post Transfer the `_valueToken` token to the `_beneficiary` address. The
```
Comment (deepesh-kn) [inline contracts/erc20-gateway/ERC20Gateway.sol]: Please update the post conditions:
```
* \post Adds a new entry in `inbox` mapping storage variable. The value is
     *       set as `true` for `messagehash_` in `inbox` mapping. The
     *       `messageHash_` is calculated by `MessageInbox.confirmMessage`.
     * \post Transfers the tokens to the `msg.sender` address as a fees.
     *       The `fees` amount is calculated by calling
     *       `ERC20GatewayBase::reward()` with parameters `gasConsumed`,
     *       `_feeGasPrice` and `_feeGasLimit`. `gasConsumed` is the approximate
     *       gas used in this transaction.
     * \post Transfer the `_amount-fees` amount of token to the `_beneficiary`
     *       address.
     * \post Update the nonces storage mapping variable by incrementing the
     *       value for `msg.sender` by one.
     * \post Emits `WithdrawIntentConfirmed` event with the `messageHash_` parameter.
```

**Review by deepesh-kn — CHANGES_REQUESTED** [review]

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 16 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Done.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 17 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Documentation Updated.

**Review by deepesh-kn — APPROVED** [review]
Approving this.
Currently, the function signature is not consistent with `withDraw` functions signature. 
We can change this later as it is giving the stack too deep error, so we can do this in another ticket.
```
function confirmWithdraw(
        uint256 _amount,
        address _beneficiary,
        uint256 _feeGasPrice,
        uint256 _feeGasLimit,
        address _utilityToken,
        address _valueToken,
        address _withdrawer,
        uint256 _blockNumber,
        bytes calldata _rlpParentNodes
    )
```

**Pull request merged**